nrf52-hal-common/src/uarte_async.rs

```diff
@@ -0,0 +1,288 @@
+            w.enable().enabled()
+        );
+
+        // Configure
+        let hardware_flow_control = pins.rts.is_some() && pins.cts.is_some();
+        uarte.config.write(|w|
+            w.hwfc().bit(hardware_flow_control)
+             .parity().variant(parity)
+        );
+
+        // Configure frequency
+        uarte.baudrate.write(|w|
+            w.baudrate().variant(baudrate)
+        );
+
+        unsafe {
+            MAYBE_CONSUMER = Some(cons)
+        };
+
+        nvic.enable(Interrupt::UARTE0_UART0);
+
+        uarte.intenset.write(|w| {
+            w.endtx().set_bit()
+        });
+
+        UarteAsync {
+            periph: uarte,
+            prod,
+        }
+    }
+
+    pub fn write_async(
+        &mut self,
+        tx_buffer: &[u8]
+    ) -> Result<(), Error> {
+        let w_grant = self.prod.grant(tx_buffer.len())
+            .map_err(|_| Error::TxBufferTooLong)?;
+
+        w_grant.buf.copy_from_slice(tx_buffer);
+
+        self.prod.commit(tx_buffer.len(), w_grant);
+
+        NVIC::pend(Interrupt::UARTE0_UART0);
+
+        Ok(())
+    }
+
+    /// Return the raw interface to the underlying UARTE peripheral
+    pub fn free(self) -> T {
+        self.periph
+    }
+}
+
+
+pub struct Pins {
+    pub rxd: P0_Pin<Output<PushPull>>,
+    pub txd: P0_Pin<Output<PushPull>>,
+    pub cts: Option<P0_Pin<Output<PushPull>>>,
+    pub rts: Option<P0_Pin<Output<PushPull>>>,
+}
+
+
+#[derive(Debug)]
+pub enum Error {
+    TxBufferTooLong,
+    RxBufferTooLong,
+    Transmit,
+    Receive,
+}
+
+#[interrupt]
+unsafe fn UARTE0_UART0() {
+    // First, check if end_tx is pending, and clear it
+    let uart = &*UARTE0::ptr();
+
+    let end_tx = uart.events_endtx.read().bits() != 0;
+
+    if end_tx {
+        uart.events_endtx.write(|w| w.bits(0));
+    }
+    let cons: &mut Consumer = MAYBE_CONSUMER.as_mut().unwrap();
+
+    use crate::uarte_async::DmaState::*;
+    let mut stat = Idle;
+
+    ::core::mem::swap(&mut stat, &mut DMA_STATUS);
+
+    match (end_tx, stat) {
+        (false, Idle) => {
+            // check for pending data, send if possible
+        }
+        (true, Idle) => {
+            // what?
+            panic!("what?");
+            // return;
+        }
+        (false, ActiveGrant((gr, sz))) => {
+            // Starting condition but already busy, let it go
+            // Place grant back!
+            DMA_STATUS = ActiveGrant((gr, sz));
+            return;
+        }
+        (true, ActiveGrant((gr, sz))) => {
+            // Complete, check for retrigger
+            let sent = uart.txd.amount.read().bits() as usize;
+            // assert!(sent < sz);
+            cons.release(min(sz, sent), gr);
+        }
+    }
+
+    let rgrant = match cons.read() {
+        Ok(gr) => gr,
+        Err(BBQError::InsufficientSize) => {
+            // No more data to send
+            return;
+        }
+        Err(_) => panic!("queue error"),
+    };
+
+    let sz = min(255, rgrant.buf.len());
```

Comment:
This should not be 255, but instead something that makes sense for both the 832 and 840